### ISSUE TITLE
Experiment Module: Update and customise the FarmOS landing page dashboard

### ISSUE THREAD

**Issue by aislinnpearson** [Rothamsted-Ecoinformatics/farm_rothamsted#679]
In order to make it easier for scientists to find their experiments, we have previously discussed the idea of updating and customising the FarmOS landing page dashboard.

My basic thinking here is that the landing page would keep the farm map at the top of the page and beneath that we would include three search fields, one each for:

1. Find a Field (search by name only; location can currently be done by the map)
2. Find a Crop assets (search by name, location, crop, variety and season) 
3. Find an Experiments (search by name, location, experiment code, study code and associated researchers)

What I am less sure about is how/ if we want to include in this a general search query as described in issue #676 

Beneath this I was thinking we could have a more customised element, where there is a data field for Proposals and Experiments the user is associated with (perhaps with a link below that takes you to your notifications page as discussed in issue xxx). Ideally we'd build this in a way that is extensible - for example, if we might in future want to add crop assets, management plans or even specific subsets of logs (e.g. those that need review) which the logged in user is responsible for.

**Comment by aislinnpearson** [comment]
Hi @paul121,

Following our call yesterday, I mocked up the below and then asked the Farm staff if it meets all their requirements:

![image](https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/assets/16349781/1d467514-ccc2-4b26-b3bc-69c83fd7f480) 

It's not obvious from my image (apologies) but we would include hyperlinks where possible - that way you can go to a study or an experiment quite easily. I also wonder if we should include 'Design name' and 'Status' along with the list of studies.

They thought everything had been included, but had a few comments:

1. **Search bar:** Please can we make sure this is as flexible as possible (e.g. not case sensitive, etc)
2. **Field Search:** Ideally this needs to include a search bar and a pick list (people don't always know the names of the fields)
3. **Experiment Search:** There was a question here about searching by experiment code, study code or name.

Given these comments, I am wondering if we make the Field, Crop and Experiment searches a pop up box - similar to the 'Select Asset' on a Log form. They can then search _or_ filter. 

The other thing I have noticed is that what people are usually looking for is Plans (which I think I am going to start referring to as Studies, because that is what they are called in the ISA framework and I think it would be good to get people used to the idea that an Experiment can have multiple studies). So for the Experiment search box, I am wondering if it should search _all_ of the Experiment module (or maybe just experiments, designs and plans/ study's) but _only_ return Experiments. That way you'd have to click on the experiment to find the plan? I am not sure. Something we should probably talk about in more detail.

**Comment by paul121** [comment]
Okay @aislinnpearson I've taken a first pass at both of these. These are both un-polished as is, but my hope is that this is close to the initial functionality we want. I can work on polishing both of these in a quick follow up. These should both be ready for testing in the demo.

Some details/questions for both of these:
- Field/Crop/Experiment search on dashboard
  - Instead of 3 or 4 text boxes this uses one, with a simple dropdown to select the type of search. This is quite rough right now - but imagine all of these being in-line in a single row, centered on the page. This wouldn't be a hard change to make, I just wanted to make sure before putting more work into it.
  - Clicking 'enter' or the 'Search' button will open results in a popup modal, which can then be used to modify & search again within the modal.
  - I went ahead and only implemented high-level Experiment entity searches (no designs, study plans, etc.)
  - For each of these you can search by any field that appears as a column in the search result.
![dashboard-search](https://github.com/user-attachments/assets/5dcfd1a5-9542-4fa7-bd40-027c4000cfe9)
![search-result](https://github.com/user-attachments/assets/4e6cae1e-0970-4cd6-99ea-062581c5243b)


- User Proposal + Studies list on dashboard
  - These are some simple table/lists to start. Right now it only shows proposals/studies that the user *created*.
  - Do you want to include proposals/studies the user is associated with? If so, which logic?
  - Also, I need some more clarification if we want to do Studies (plans) or Experiments!
  - Do we want to only include some statuses? Display anything that is active/pending, but hide archived?
  - How would we like to handle lots of results? I can make a scrolling area so the top ~5 are shown but the rest need to be scrolled.
  
![Screenshot from 2024-08-04 23-51-35](https://github.com/user-attachments/assets/8a784499-8735-4ac9-8785-2543ae462687)

**Comment by paul121** [comment]
Oh also - right now this still includes the existing dashboard info. But I can remove these items!

**Comment by aislinnpearson** [comment]
Hey @paul121,

Thanks for this. It's definitely getting there. I had a quick look and here are some initial thoughts: 

**Search functionality:**

> Instead of 3 or 4 text boxes this uses one, with a simple dropdown to select the type of search. This is quite rough right now - but imagine all of these being in-line in a single row, centered on the page. This wouldn't be a hard change to make, I just wanted to make sure before putting more work into it.

I like the single search option - it will definitely look much better aligned (the box currently takes up a lot of my wide screen so unless you know they are there you don't see your experiments and plans). A couple other observations:

1. instead of '_entity type_' can we make the title '_search by_' or something a bit more intuitive. I am not sure the scientists always know what an entity in FarmOS is. 
2. Ideally we'd also have four options in the box: _field_, _crop asset_, _experiment_ and _any_ or _all_, but maybe that is more broadly linked to issue #676?

> Clicking 'enter' or the 'Search' button will open results in a popup modal, which can then be used to modify & search again within the modal.

This was a bit counter intuitive because the search opens a pop up box, which can then also be used for multiple searches. I wonder if it could be a single search button that opens the pop up box (without having to do the initial search? Although that would be one more click...), but am I right in saying this was the more complicated functionality? 

> I went ahead and only implemented high-level Experiment entity searches (no designs, study plans, etc.)

What I was hoping with the experiments is that would search any field in the experiment, design or plan. So, for example, if I select experiment and type in the study code '2020/R/BK/1' it would bring up any plans _and_ experiments where that code is listed. Maybe the results could be categorised by experiments, designs and study plans... not sure what is possible, so let me know if this idea needs rethinking

> For each of these you can search by any field that appears as a column in the search result. 

Would it be possible to make it any field within the entity? For example, if I type 2023 it brings up any plans, designs and experiments with 2023 as the start or end year? This would apply to whatever entity is selected (i.e. if I select crop assets it searches all data fields on the crop assets). Again, let me know what's possible.

> Oh also - right now this still includes the existing dashboard info. But I can remove these items!

Sounds good. Looking at the four available (upcoming tasks, late tasks, active plans and metrics) I'd say the first two could easily be replaced by the farm calendar, but that can be a separate issue (see here: #730). Also, I like the active plans - those are useful for the farm team if they were experiment plans, but I think let's wait to implement the plan module first. I also quite like the idea of a metrics tab (#731). If I was to keep anything there it would probably be flagged issues, but maybe that needs to be another feature request (e.g. flagged assets, logs related to the person logged in and/or all if you are the farm manager or data admin). (#732)

The last thing I noticed is that, although it is not caps sensitive, you need the exact spelling. It doesn't do fuzzy matches. E.g. if you have a typo and spell '2020/R/BK1' instead of '2020/R/B**/**K1' (a real live mistake I found for one of the experiments last week) you get no results, and might therefore assume it doesn't exist even though it does. I realise this is the initial testing phase, and we agreed this as an intermediary option to get us part of the way there, but I definitely think it is going to be something we need to keep an eye on. 

What would be good to document is the limits of the search functionality though if those are written anywhere, and we can link to them?

**Links to Proposals and Experiments**

> These are some simple table/lists to start. Right now it only shows proposals/studies that the user created. 
> Do you want to include proposals/studies the user is associated with? If so, which logic?

The logic should please be:
- **Proposals:** if you are named on a proposal, it shows in your list of proposals
- **Study Plans:** if you are named on an experiment _or_ a design associated with a study plan, it shows up in your list

> Also, I need some more clarification if we want to do Studies (plans) or Experiments!

Studies please. The list of headings you have is almost perfect. Ideally it would be:

- Status (of the study)
- Study Name (as a hyperlink)
- Study period id
- Experiment (as a hyperlink)
- Design (as a hyperlink) 
- Location (as a hyperlink)

My thinking here (and I may be very wrong) is that I need to scientist to start separating out the concept of a study from an experiment. I hope this will help me achieve that, but only time will tell. 

> Do we want to only include some statuses? Display anything that is active/pending, but hide archived?

An excellent point. I agreed it should be everything _except_ archived. The ideal solution would be a button that says 'show archived', but let me know if I need to create a separate issue for that (I suspect I probably do?)

> How would we like to handle lots of results? I can make a scrolling area so the top ~5 are shown but the rest need to be scrolled.

Another excellent point... it would either be scroll or pop up I guess? If it is scroll, then I think we need to lay the proposals and the studies out underneath each other rather than next to each other? 

I'm really keen to release this as a 'tell me what you think' issue for feedback from staff, so maybe we could even try a couple of options and then I ask them to tell me what they prefer? 

Sorry @paul121 - I realis that is a lot of comments, but I really love the progress. It's going to be such a nice improvement when it's done, even if it isn't perfect on the first release.

**Comment by aislinnpearson** [comment]
For experiments only:
1. Experiments
2. Designs
3. Study Plans

For all:
1. Assets (with type listed, e.g. boundary or field)
2. Experiment (with name and type listed)
3. Design
4. Study Plans
5. Logs (with type listed)

**Comment by paul121** [comment]
Okay, I've made the style changes so the search bar is inline and added placeholder helper text. I also updated the columns for the "My Studies" dashboard item.

![Screenshot from 2024-08-05 21-50-09](https://github.com/user-attachments/assets/e136cbc2-232d-494a-a0cd-cd3ad018d4a2)

@aislinnpearson I don't think I'll have time to do the "Grouped search results by entity type" like we discussed. We could deploy this as-is or wait till I'm back? I'll update the demo site now.

**Comment by aislinnpearson** [comment]
That looks great @paul121!

Let's wait until you are back. I am happy to test variations in visualisation, but I'd like the functionality to be as good as we can get it. Otherwise I think there will just be misunderstandings and complaints.

**Comment by aislinnpearson** [comment]
The tick box list as promised @paul121:

**Before the next release:**
- [x] Experiment search: expand to include all fields for experiments, designs and plans. Display Experiments, Designs and Plans as sub-headings in the search.
- [x] Only display a proposal on the dashboard if the person logged in is named on that proposal or experiment (I think the current logic is named _or_ created, but sometimes Leo, the data stewards or I will create entitles for other people).

Once these two are addressed, I am happy to release a version for feedback, so the remaining list (in order of priority) are only nice to have before the next release, depending on what you have time for. We will probably want to expand on this anyway once we have feedback.

**Remaining items:**
- [x] Change the layout of the 'my proposals' and 'my studies' boxes so that they are stacked on top of each other (plans on top of studies), and display in the same way as the current filter functionality. This is to make it easier to include long names for experiments, designs, plans, etc.
- ~~[ ] add search all to the list of options (issue #676)~~ **see comment below**
- [ ] Add a 'show archive' button (needs discussion as we might want to show archived, cancelled and completed) - **see comment below**
- [ ] Remove additional dashboard information (upcoming tasks, late tasks, metrics, active plans)
- [x] Make the boxes collapsed by default (if possible)
- [x] Change the order, so studies come first

**Comment by paul121** [comment]
Okay @aislinnpearson I've completed the first 3 checkboxes above and updated the demo - let's see how this goes and save others for the next milestone with future feedback!

**Comment by aislinnpearson** [comment]
@mstenta @paul121 - this looks good to me. @mstenta is there any chance we could get it pushed before Monday? We have someone visiting us from CSIRO and I was hoping to show them how we are using FarmOS...

**Comment by mstenta** [comment]
I should be able to do that, yes. @paul121 can you publish a new release when you're ready? I can try to get it out over the weekend.

**Comment by aislinnpearson** [comment]
**Feedback:**

- **User feedback from Loreto (a data steward):** None of the proposals or experiments she is named on appear on her dashboard. For the proposals I think this is because she is named as a data steward (we would have the same issue for statisticians). For the studies I am struggling more to understand what is happening. Both she and I are named as Data Curators, and all my experiments appear. The only difference I can see is that she is not a Farm Admin or a Reviewer (I am). Loreto's Researcher [Profile](https://rothamstedfarm.farmos.net/rothamsted/researcher/38, the [list of her proposals](https://rothamstedfarm.farmos.net/rothamsted/proposal/all?experiment_category=All&status=All&name=&planting_year=&harvest_year=&statistician_target_id=&data_steward_target_id=Loreto+Pino+Chandia+%2838%29&contact_target_id=) and the[ list of her experiments](https://rothamstedfarm.farmos.net/rothamsted/experiment?category=All&status=All&name=&abbreviation=&code_value=&researcher_target_id=Loreto+Pino+Chandia+%2838%29).
- **General observation**: 'Experiments' shoud be named 'Research and Experiments' and the search expanded to include Programs and People.

**Comment by aislinnpearson** [comment]
Hey @paul121,

I've updated https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/issues/679#issuecomment-2288727126 with the additional boxes as promised. 

For the 'show archived' I think:
**Proposals** should show 'Draft, Submitted, Approved, Rejected, Planned' by default with a button for 'show completed, cancelled and archived'. 
**Plans** should show 'Active, Requested and Planning' by default, again with a button for 'show completed, cancelled and archived'. 

I am not thrilled with my decision. Please feel free to contribute if you can see a better solution. We might want to change this when it comes to reviewing the Archiving process (see #716), and I added this comment to that parent tracking issue so we don't loose sight of it when dealing with those processes/ issues.

**Comment by paul121** [comment]
Okay @aislinnpearson I made the easy changes, making these collapsible and changing the order. A couple questions:

> Remove additional dashboard information (upcoming tasks, late tasks, metrics, active plans)

Looking into this I've realized it might be a bit harder than I had hoped. I'd like to contribute a feature to farmOS core to make this easier. @aislinnpearson is it OK if we wait on this?

> For the 'show archived' I think:
> I am not thrilled with my decision. Please feel free to contribute if you can see a better solution.

Ah, I might have been confused. I thought we were talking about adding a 'Show archived' checkbox to the Search form, not to the dashboard studies and proposals. Adding this to the dashboard is harder because it is static, unlike the search which is inherently dynamic. But I understand the desire to only show 'active' statuses on the dashboard. Maybe we can make this change to only show 'active' statuses, but not add the checkbox? I've updated the result text to read '5 active proposals' or '5 active studies'. I know "active" isn't the perfect term here, but do you think that is sufficient?

![dashboard-active-entities](https://github.com/user-attachments/assets/3b7f4081-75f8-4e72-bc8b-8be436f5a06e)

**Comment by aislinnpearson** [comment]
Hey @paul121,

> Looking into this I've realized it might be a bit harder than I had hoped. I'd like to contribute a feature to farmOS core to make this easier. @aislinnpearson is it OK if we wait on this?

100%. I'll made a separate issue for this so we can keep track of it. See here: #748 

> Maybe we can make this change to only show 'active' statuses, but not add the checkbox? I've updated the result text to read '5 active proposals' or '5 active studies'. I know "active" isn't the perfect term here, but do you think that is sufficient?

I am not sure I fully understand... can we discuss it on the call today? As an aside, I asked one of our data stewards to show you how long the lists can get (so even if you only show the top 5, it's still very long). Even I needed this for perspective - my list is not this long!!

![Picture1](https://github.com/user-attachments/assets/ce84d408-c56e-4444-915d-59abe0162d7c)

**Comment by paul121** [comment]
> As an aside, I asked one of our data stewards to show you how long the lists can get (so even if you only show the top 5, it's still very long). Even I needed this for perspective - my list is not this long!!

Fixed the issue with scrolling! This issue should be done now. Will tag in a new 2.23.1 release.
